### PR TITLE
Reduce types output

### LINKED ISSUE #8613: Reduce size of typescript input/output definitions files
Team, looks like typescript definitions embedded in one single package (azure-native) drains tons of resources on the development machine, is it possible to divide the package into multi-package / plugin per namespace? Something like: @pulumi/azure-native-network . Another possible solution would be to not create one giant input types definition file and make it per namespace. Thanks

azure-native 2021-05-19 07-37-34

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
# Description

This is a draft PR for #8613.

Fixes #8613 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/nodejs/gen.go [hunk 1]
```diff
@@ -1905,12 +2008,13 @@ func (mod *modContext) gen(fs fs) error {
 	// Nested types
 	// Importing enums always imports inputs and outputs, so if we have enums we generate inputs and outputs
 	if len(mod.types) > 0 || (mod.pkg.Language["nodejs"].(NodePackageInfo).ContainsEnums && mod.mod == "types") {
-		input, output, err := mod.genTypes()
+		files, err := mod.genTypes()
 		if err != nil {
 			return err
 		}
-		fs.add(path.Join(modDir, "input.ts"), []byte(input))
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:1912]: At first I was thinking it's breaking but the new code seems to gen input.ts too. Interesting.

File: pkg/codegen/nodejs/gen.go [hunk 2]
```diff
@@ -1617,6 +1635,133 @@ type namespace struct {
...
+	} else {
+		filename = path.Join(dirRoot, "outputs.ts")
+	}
+	fmt.Printf("Generating namespace into IO file: %s\n", filename)
+	var file = newIOFile(filename)
+	// We start every file with the header information.
+	ctx.mod.genHeader(file.writer(), ctx.mod.sdkImports(true, false), ctx.externalImports, ctx.imports)
+	// We want to organize the items in the source file by alphabetical order.
+	sort.Slice(ns.types, func(i, j int) bool {
+		return tokenToName(ns.types[i].Token) < tokenToName(ns.types[j].Token)
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:1692]: This is not a total order, leading to non-determinism. This sort func has been fixed on `master` I think.

File: pkg/codegen/nodejs/gen.go [hunk 3]
```diff
@@ -1617,6 +1635,133 @@ type namespace struct {
...
+	// We have successfully written all types at this level.
+	// Next, we recurse to generate the files at the next directory level.
+	sort.Slice(ns.children, func(i, j int) bool {
+		return ns.children[i].name < ns.children[j].name
+	})
+	for i, child := range ns.children {
+		// At this level, we export any nested definitions from
+		// the next level.
+		fmt.Fprintf(file.writer(), "import * as childname(%s) from \"childname(%s)\";\n", child.name, child.name)
+		fmt.Fprintf(file.writer(), "export type child(%s);", child.name)
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:1722]: These are the re-exports? Interesting. 

I'm a bit confused how the childname and child() render.

Including the results of applying this codegen change to examples (PULUMI_ACCEPT=1 go test -test.v TestGeneratePackage) in the PR can help illusrate.

Also I would be vary of `export type`.  Syntactically consider re-exports https://www.typescriptlang.org/docs/handbook/modules.html#re-exports

Semantically consider the multiple levels here. `export type` will only export Type-level bindings. However I believe we might have value-level bindings in there as well? For example a class declaration has Type-level and Value-level bindings. There are also Namespace-level bindings. It's confusing.

I've just browsed through aws and tls types/{input,output}.ts. It would seem that most of the exports are exporting interfaces so `export type` is fine. There's a few places that do `export namespace`.

**Comment by t0yv0** [comment]
Looking promising!

File: pkg/codegen/nodejs/gen.go (hunk 3 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1722]: Sorry, I tried to indicate previously (but it's been awhile and I should have written it down in line!) that this is the major WIP part of the PR. It's just absolutely not the right path. `childname` is a placeholder I added to help me trace the code.

> I've just browsed through aws and tls types/{input,output}.ts. It would seem that most of the exports are exporting interfaces so export type is fine. There's a few places that do export namespace.

I think each `export namespace` translates 1:1 with a new file created as part of this PR. When we create a new `namespace`, we're really trying to create a new layer of nesting. See the images attached below for how this translates.

Forgive me if I'm not explaining this well, I can my articulation is suffering from a lack of clarity.

directories
namespaces

Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1722]: Trying to take your suggestion to use the `export * as x from "x"` syntax (from your linked resource) because you're right about the value-bindings. Right now it's looking like value bindings are still missing from those exports, which is strange, since they shouldn't be.

File: pkg/codegen/nodejs/gen.go (hunk 2 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1709]: I rebased off of `master` and pulled in the change, which I noticed is indeed on `master`.

**Comment by RobbieMcKinstry** [comment]
**Update:** I'm successfully building AWS-Native with these changes. 
I also rebased off of master to get back up to date. I still need to cleanup the code (lots of Println statements) and update the tests.

**Comment by t0yv0** [comment]
@RobbieMcKinstry did you get a chance to look at CI failures?

For example this one:

TestGeneratePackageTwo/plain-object-defaults/nodejs/compile 

Seems to be something off with finding the `utilities` module

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts:2]: Special case for empty file to not emit empty modules? WDYT?

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/output.ts [hunk 5]
```diff
@@ -5,33 +5,4 @@ import * as pulumi from "@pulumi/pulumi";
 import * as inputs from "../types/input";
 import * as outputs from "../types/output";
 
-export namespace documentdb {
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/output.ts:8]: Lovely. So the `export *` replacement does the same thing as before? This looks right.

File: pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod1/input.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * A test for namespaces (mod 1)
+ */
+export interface TypArgs {
+    val?: pulumi.Input<string>;
+}
+/**
+ * typArgsProvideDefaults sets the appropriate defaults for TypArgs
+ */
+export function typArgsProvideDefaults(val: TypArgs): TypArgs {
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod1/input.ts:19]: This is a concrete example of a non-type export in these inputs files. If we're still in doubt if re-export works as intended, we could check calling this engaging the re-export.

**Comment by RobbieMcKinstry** [comment]
I spent 3.5 hours chasing down a bug around inconsistent module import paths. I fixed it! I've got four failing tests left to get to:
- [ ] provider-config-schema
- [ ] plain-object-disable-defaults
- [ ] plain-object-defaults
- [ ] plain-and-default

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts (hunk 4 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts:2]: That might be easier in a follow-on PR. Right now there are a lot of edge cases I'm coming up against regarding this same pattern. Before we merge, I can file an issue to handle this as a special case.
Suppose we implement a special case to omit an empty file. If either input or output module is empty, then the index file (which is generated separately) has to know that too, so that it doesn't generate an import for a file that isn't there.

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/output.ts (hunk 5 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/output.ts:8]: Yeah, `export * as ns from "file";` creates a namespace and re-exports everything exported by `file`. I read about it in the [v3.8 release notes](https://devblogs.microsoft.com/typescript/announcing-typescript-3-8/#export-star-as-namespace-syntax):

```typescript
export * as utilities from "./utilities";
```

is a syntactic sugar for:

```typescript
import * as utilities from "./utilities";
export { utilities };
```

File: pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod1/input.ts (hunk 6 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod1/input.ts:19]: Ah, the test `plain-object-defaults` does exactly this. I had a name error on this very type until very recently. 
I also removed the `export type` syntax in favor of `export * as ns` so I think its a nonissue.

File: pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod2/index.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+export * from "./input";
+export * from "./output";
+export * as inputs from "./input";
+export * as outputs from "./output";
```
Comment (RobbieMcKinstry) [inline pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod2/index.ts:7]: @t0yv0 This is the only major change of note since your last review.

Each file matching the format `types/{*}/index.ts` will export names two ways. I'll explain.

Suppose you have a module `/types/foo/index.ts`, and an input named `MyInput`.

1. You can import `MyInput` on the index itself: 

```typescript
import * as foo from "types/foo";
const input: MyInput = null;
```

2. You can also import it from inside the `input` namespace on the index:

```typescript
import * as foo from "types/foo";
const input: foo.input.MyInput = null;
import * as fooInputs from "types/foo/input";
const input2: fooInputs.MyInput = null;
```
It's kind of gross to expose the name twice. However, as we know from #10442 there are several locations where imports aren't smart enough to identify the exact name of the required import. I'm aiming to fix that when I tackle #10442.

Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/plain-object-defaults/nodejs/types/mod2/index.ts]: Nevermind this comment. I removed the second namespace, since it's not required for backwards compatibility and makes the API really confusing.

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts (hunk 4 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/input.ts:2]: Actually, I suppose I can't open the issue until after this is merged, otherwise I'd be opening an issue about a problem that doesn't exist yet lol

**Comment by RobbieMcKinstry** [comment]
All codegen tests pass. Downstream codegen tests fail. I can't repro locally. However, I think the problem was a conflict in namespace declarations. I was exporting child modules from both input.ts and output.ts so there was a name conflict.

**Comment by RobbieMcKinstry** [comment]
🚀 Yay! I got tests to pass! ✨ 

Wow, this was incredibly difficult. Definitely the hardest PR I've worked on at Pulumi.
But it's now ready for final review and/or approval!

**Comment by squaremo** [comment]
A substantial piece of work, nice one @RobbieMcKinstry! I really appreciate that you kept commits to the point, and rebased to keep the history clean (there's a few WIP and reruns that you could `rebase -i` out) 🙏

File: pkg/codegen/nodejs/gen.go [hunk 8]
```diff
@@ -1463,14 +1463,14 @@ func (mod *modContext) genHeader(w io.Writer, imports []string, externalImports
 		sort.Strings(modules)
 
 		for _, module := range modules {
-			fmt.Fprintf(w, "import {")
+			fmt.Fprintf(w, "import { ")
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:1466]: The addition of a space here is for aesthetic/ style guide? Consider being kind to the providers team, any change currently trips diffs for us to review so if not strictly necessary err on omitting, or separate PR/change.

File: pkg/codegen/nodejs/gen.go [hunk 9]
```diff
@@ -1617,6 +1684,223 @@ type namespace struct {
...
+		var indexFile = ns.genIndexFile(ctx, dirRoot)
+		files = append(files, indexFile)
+	}
+	return files, nil
+}
+
+// sortItems will sort each of the internal slices of this object.
+func (ns *namespace) sortItems() {
+	sort.Slice(ns.types, func(i, j int) bool {
+		return tokenToName(ns.types[i].Token) < tokenToName(ns.types[j].Token)
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:1779]: This sort is possibly non-deterministic again (see my earlier comment). What's the the of `ns.types`?

File: pkg/codegen/nodejs/gen.go [hunk 10]
```diff
@@ -2294,6 +2587,7 @@ func genTypeScriptProjectFile(info NodePackageInfo, files fs) string {
 
 	fmt.Fprintf(w, `{
     "compilerOptions": {
+		"baseUrl": ".",
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:2590]: Could we possibly fix indent here to be consistent (tabs vs spaces, width).

File: pkg/codegen/nodejs/gen.go [hunk 11]
```diff
@@ -2304,6 +2598,9 @@ func genTypeScriptProjectFile(info NodePackageInfo, files fs) string {
         "experimentalDecorators": true,
         "noFallthroughCasesInSwitch": true,
         "forceConsistentCasingInFileNames": true,
+		"paths": {
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:2601]: Curious what does this do?

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/output.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+import * as pulumi from "@pulumi/pulumi";
+import * as inputs from "../../types/input";
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/output.ts:5]: Cool! So our next difficult task will be to make these imports here more precise 😬

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts [hunk 13]
```diff
@@ -4,6 +4,6 @@
 import * as pulumi from "@pulumi/pulumi";
 import * as inputs from "../types/input";
 import * as outputs from "../types/output";
+import * as utilities from "../utilities";
 
-export namespace documentdb {
-}
+export * as documentdb from "./documentdb/input";
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts:9]: Hm, interesting! So it only re-exports /input.

**Comment by t0yv0** [comment]
@RobbieMcKinstry I didn't see any re-exports concerning enum but perhaps that's ok since those are not moving?

Elegant treatment of per-input/per-output reexports! I guess this way TypeScript namespace merging gives the exact desired result without codeine having to track the names of the re-exported members. Very neat!

**Review by t0yv0 — APPROVED** [review]
Looking good modulo some nits/questions! Thank you! Appreciate another review from platform team.

**Comment by RobbieMcKinstry** [comment]
bors cancel

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/output.ts (hunk 12 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/documentdb/output.ts:5]: Yes! Indeed it is!

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts (hunk 13 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts:9]: Yeah, that was what I was getting hung up on for the longest time.
I tried:

```typescript
export * as documentdb from "./documentdb";
```

which would export everything in the index file, but that had some knock-on effects to preserve the API:

1. If I tried to have the index file re-export both inputs and outputs in the _same_ namespace, this would introduce a name collision between `documentdb/inputs.ts::Foo: Input<T>` and `documentdb/outputs.ts::Foo: Output<T>` (or something to this effect).
2. If I tried to have the index file re-export `inputs.ts as inputs` and `outputs.ts as outputs`, then I would break backward-compat. You'd have to import things with this syntax `import "types/inputs/documentdb/inputs/..."`.

This approach agglomerates the nested input items in the parent's namespace, and same with outputs. This preserves backward-compat with what we have today.

**Comment by RobbieMcKinstry** [comment]
> I didn't see any re-exports concerning enum but perhaps that's ok since those are not moving?

Yeah, I didn't touch enums at all. Those could/should be a separate issue if they get too large. For our biggest packages, there still aren't an insane number of enums. Long term, I think we should move the enums into the same structure (or re-export them from the folder most closely related to them). I'm unsure why they're separate. They should be in the index of each directory e.g. `types/ec2/index.ts` should contain the enums related to the `ec2` module. Do you agree? I could be missing something.

File: pkg/codegen/nodejs/gen.go [hunk 14]
```diff
@@ -1617,6 +1684,223 @@ type namespace struct {
...
+
+// genIndexTypes generates an index.ts file for this directory. It must be called
+// only once. It exports the files defined at this level in input.ts and output.ts,
+// and it exposes the types in all submodules one level down.
+func (ns *namespace) genIndexFile(ctx *ioContext, dirRoot string) *ioFile {
+	var indexPath = path.Join(dirRoot, "index.ts")
+	var file = newIOFile(indexPath)
+	ctx.mod.genHeader(file.writer(), nil, nil, nil)
+	// Export the types defined at the current level.
+	// fmt.Fprintf(file.writer(), "export * from \"./input\";\n")
```
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1861]: Yes!

File: pkg/codegen/nodejs/gen.go (hunk 9 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1779]: Good catch! Thanks!

File: pkg/codegen/nodejs/gen.go (hunk 8 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1466]: I'll revert. Yeah, I bet `prettier` probably adds a space when it processes the code (whenever that is). It's not at all necessary; I was just being pedantic. If it's an inconvenience at all, then it's not worth it.

**Comment by RobbieMcKinstry** [comment]
> A substantial piece of work, nice one @RobbieMcKinstry! I really appreciate that you kept commits to the point, and rebased to keep the history clean (there's a few WIP and reruns that you could `rebase -i` out) 🙏

Will do! I'll rebase again when I address Anton's feedback. :) Thanks for taking a look Michael!

File: pkg/codegen/nodejs/gen.go [hunk 15]
```diff
@@ -1537,77 +1537,141 @@ func (mod *modContext) genConfig(w io.Writer, variables []*schema.Property) erro
...
 	}
 
 	return imports
 }
 
-func (mod *modContext) utilitiesImport() string {
-	relRoot := mod.getRelativePath()
+func (mod *modContext) utilitiesImport(relRoot string) string {
 	return fmt.Sprintf("import * as utilities from \"%s/utilities\";", relRoot)
 }
```
Comment (kpitzen) [inline pkg/codegen/nodejs/gen.go:1587]: Is there a benefit to keeping this a method?  Or do we just wish to preserve the association between modContext and this function?

File: pkg/codegen/nodejs/gen.go [hunk 16]
```diff
@@ -1617,6 +1681,221 @@ type namespace struct {
...
+func (ctx *ioContext) filetype() string {
+	if ctx.input {
+		return "input"
+	}
+	return "output"
+}
+
+// intoIOFiles converts this namespace into one or more files.
+// It recursively builds one file for each node in the tree.
+// If input=true, then it builds input types. Otherwise, it
```
Comment (kpitzen) [inline pkg/codegen/nodejs/gen.go:1727]:
```suggestion
// If ctx.input=true, then it builds input types. Otherwise, it
```

**Review by kpitzen — APPROVED** [review]
I have nothing but the tiniest of nits for this, neither of which I would consider blocking.  OTOH, I _love_ the attention to detail paid to comments and cleanliness in this PR.  Fantastic work.

File: pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json [hunk 17]
```diff
@@ -10,6 +11,9 @@
         "experimentalDecorators": true,
         "noFallthroughCasesInSwitch": true,
         "forceConsistentCasingInFileNames": true,
+		"paths": {
+			"@/*": ["*"]
+		},
```
Comment (AaronFriel) [inline pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: The tsconfig changes seem like no-ops. Are they? 

`baseUrl` and `paths` don't appear used, as I don't see any absolute paths, and they introduce the potential for name collisions between absolute paths and the names of dependencies. 

As we're generating all of these files, we shouldn't need path mangling: we can generate correct paths from the get-go:


baseURL: `import * as x from "hello"` could resolve to `./hello` or `node_modules/hello`

paths: `import * as x from "@/foo"` doesn't appear used.

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts (hunk 13 above)
Comment (t0yv0) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/types/input.ts:9]: This Is the key in the PR. Brilliant.

**Comment by t0yv0** [comment]
Don't actually care about enums just checking that we don't break bw-compat on those, sounds like we don't! They're not large enough to be worth optimizing.

**Review by AaronFriel — COMMENTED** [review]
Question about tsconfig changes introduced, I see a related thread but I looked & did not see them used.

File: pkg/codegen/nodejs/gen.go (hunk 15 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/nodejs/gen.go:1587]: Not really. No reason it can't be a function.

File: pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json (hunk 17 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: @AaronFriel It's not a no-op. It's used [here](https://github.com/pulumi/pulumi/pull/10831/files#diff-f7917202f932e91c5a03a811977f165efebbc5c006159c0a8865e3a93713048cR1655).

TL;DR: The mangling will be removed in https://github.com/pulumi/pulumi/issues/8613 and having absolute paths makes that much easier IMO.

> As we're generating all of these files, we shouldn't need path mangling: we can generate correct paths, rather than relying on aliasing/rewrite rules in tsconfig.

I assume you're referring to [this snippet](https://github.com/pulumi/pulumi/pull/10831/files#diff-f7917202f932e91c5a03a811977f165efebbc5c006159c0a8865e3a93713048cR1647-R1659), which translates paths relative to `/types` to a top-level path.

You're right, it is technically possible to track the correct relative path and use that instead. I would have used that approach but it became probability invasive. Here's why:

I found that the existing logic for generating relative paths incorrect. `mod.getRelativePath()` operates using the `mod.mod` field, which is always a single path, so the function will always return `..` no matter what mod it's operating on. Further, I had a _rough_ time adapting the logic to the nested types directories. Many functions assume that the code generated is at most two levels removed from the parent at any time, which is not true after this PR. Lots of high level functions call `getRelativePath` deeply in their callstack.
One workaround is to adopt a recursive approach where we carry the context of how deep we are between function calls. In general, this would need to adjust some number of functions to accept a `parentDir` parameter to track how deep we are in the file hierarchy (or, since we know the depth of the filesystem is <= 2, do something less DRY). This parameter needs to be threaded to a number of existing methods that don't already have it, and any calls to those methods need to supply a default value of some kind. 

I actually implemented this on `intoIOFiles` before realizing I could use `paths` (it's still in there now). The knock-on effects of threading this through `intoIOFiles`:

1. Must add the parameter to `genOwnedTypes`.
2. ...which means I have to add the parameter to `sdk.ImportsWithPath`
3. ...which means threading it through `getRelativePath` because the original logic was no longer valid. 
4. Same with `genNestedTypes` because it is corecursive with `intoIOFiles`.
5. Same with `genIndexFile`

I think, ultimately, the architecture using an absolute path will look similar to what you have in mind here. As part of https://github.com/pulumi/pulumi/issues/8613 we need to build a map between type decls and their defn dependencies. If we use relative paths, then we have to recompute the map for each subdirectory. If we use absolute paths, we only need to compute it once (or once per mod context, depending on which end of the elephant you stare at).

I actually think mapping the package root to `@` via `paths` can dramatically simplify our code, both by letting us refactor existing code and also making future changes easier.
For existing code, this lets us rip out `getRelativePath`, which as I mentioned is vacuously correct. For future code, it enabled a cleaner approach to implementing https://github.com/pulumi/pulumi/issues/8613.

Aside: I also think for most of our imports abs paths will be easier to read in the generated code than relative paths. Most of the imports are `@/s3` or `@/ec2`. With relative paths, you can have both `../../s3` and `../s3`, and they point to the same file.

Comment (AaronFriel) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: I hear you, though I looked through (twice now) all of the files generated by tests and no lines change match either of these patterns. If the PR introduces functionality not tested, I'm not sure how I'm to evaluate it - would this appear if I used this branch to run sdkgen against a provider?

Comment (AaronFriel) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: FWIW, if I use a tool instead of eyeballing it: `git diff origin/master..HEAD | rg 'from "@'`  and likewise `from "\w+`.

**Comment by RobbieMcKinstry** [comment]
More to Friel's point, I'd really love to see our codegen backends turn into more traditional multi-pass compilers. Would be nice to make as many passes as we need to generate optimal code with the least tech debt.

File: pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json (hunk 17 above)
Comment (AaronFriel) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: I don't disagree with using absolute paths if it simplifies codegen & your case with @t0yv0 was compelling, apologies if my comment lead you to believe I'm opposed in general.

Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: I thought you were asking if it was used in the generator anywhere, I gotcha now! :) 

> I looked through (twice now) all of the files generated by tests and no lines change match either of these patterns.

Now that you mention it, I think you're right that the codegen tests don't exercise that code path. I introduced it as a fix failing downstream codegen CI jobs. Those jobs exercise the codepath in `pulumi-aws` at the least (not sure where else).

It occurs whenever an input or output type needs to reference a resource type.

Here's an example from `pulumi-aws/types/s3/output.ts`, which imports `pulumi-aws/s3`:
Screen Shot 2022-11-09 at 6 28 20 PM

Do you want me to add a codegen test for this case?

> apologies if my comment lead you to believe I'm opposed in general.

No need to apologize! I appreciate the thorough code review. I wanted to make my case for the absolute imports anyway because I want to get feedback from yinz guys on whether its a good idea or not :) This is precisely what I want :)

File: pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/tsconfig.json [hunk 18]
```diff
@@ -10,13 +11,19 @@
         "experimentalDecorators": true,
         "noFallthroughCasesInSwitch": true,
         "forceConsistentCasingInFileNames": true,
+		"paths": {
+            "@/*": ["*"]
+		},
```
Comment (justinvp) [inline pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/tsconfig.json:16]: Nit: It looks like this is using tabs rather than spaces for the indentation. Mind fixing it to consistently use spaces?

Comment (RobbieMcKinstry) [reply pkg/codegen/testing/test/testdata/azure-native-nested-types/nodejs/tsconfig.json:16]: Damn, I thought I fixed that. Thanks Justin!

File: pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json (hunk 17 above)
Comment (AaronFriel) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: Yes, definitely to adding or modifying an existing codegen test to help us catch a regression.

And a nit on the tsconfig file format: it's using alternating tabs and spaces. :) 


```txt
        "forceConsistentCasingInFileNames": true,
		"paths": {
			"@/*": ["*"]
		},
```

Comment (AaronFriel) [reply pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json:16]: Ah, Justin already commented on that.

**Comment by AaronFriel** [comment]
@RobbieMcKinstry We do have some passes already - hoisting optionals in Go or rewriting property traversals to applies or proxied accesses in Python & TypeScript - but you & other folks here know a lot more about compiler design than I do. Some of it occurs up front, some of it is messier & involves inline lowering of expressions. It's inconsistent when & where it occurs, which is probably a source of at least a few bugs.

File: pkg/codegen/testing/test/testdata/type-references-resource/nodejs/tsconfig.json [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        "target": "es2016",
+        "module": "commonjs",
+        "moduleResolution": "node",
+        "declaration": true,
+        "sourceMap": true,
+        "stripInternal": true,
+        "experimentalDecorators": true,
+        "noFallthroughCasesInSwitch": true,
+        "forceConsistentCasingInFileNames": true,
+		"paths": {
```
Comment (AaronFriel) [inline pkg/codegen/testing/test/testdata/type-references-resource/nodejs/tsconfig.json]: Indentation?

**Review by AaronFriel — APPROVED** [review]
Appreciate adding the test! 

Indentation on tsconfig still seems off, otherwise LGTM!

**Comment by RobbieMcKinstry** [comment]
> Appreciate adding the test!
> 
> Indentation on tsconfig still seems off, otherwise LGTM!

I think I got it now! Whitespace is hard! XD
https://github.com/pulumi/pulumi/pull/10831/files#diff-f7917202f932e91c5a03a811977f165efebbc5c006159c0a8865e3a93713048cR2588-R2590

**Comment by RobbieMcKinstry** [comment]
bors retry

**Comment by RobbieMcKinstry** [comment]
bors retry
